nrf: Replace custom-HAL with nrfx-HAL

glennrub opened the pull request:
Doing an early PR to show what is comming...

This patch series is currently Work-In-Progress. The goal is to replace the custom HAL layer in the port with Nordic Semiconductor's new open source friendly HAL layer (BSD-3) which also supports both nrf51 and nrf52. 
Ref: https://github.com/NordicSemiconductor/nrfx
 
Currently status is that Pin/LEDs seems to work OK, and UART REPL.

The old hal/ and device/ folder will be removed in the end.

glennrub committed:
ports/nrf: Initial commit for moving hal to Nordic Semiconductor BSD-3 licensed nrfx-hal.

glennrub committed:
ports/nrf: Adding nrfx, Nordic Semiconductor BSD-3 hal layer, as git submodule checked out at lib/nrfx.

glennrub committed:
ports/nrf/modules/machine/uart: Fixing bug which set hwfc to parity excluded, always resulting in no flow control, hence corrupted output. Also adding an extra loop on uart_tx_char to prevent any tx when any ongoing tx is in progress.

glennrub committed:
ports/nrf/i2c: Moving I2C over to nrfx driver.

aykevl commented on ports/nrf/mphalport.c:
```diff
@@ -77,3 +77,87 @@ void mp_hal_stdout_tx_strn_cooked(const char *str, mp_uint_t len) {
+        " NOP\n"
+        " NOP\n"
+        " NOP\n"
+        " NOP\n"
```
I think this list of NOPs can be made more compact by using a small for loop, or even jumping to the next instruction. For example `b . + 2` takes up 3 cycles on an M0 processor instead of 1 for a `nop` (thus using 1/3 of the instructions). Sadly the nRF52 is less predictable so might need these nops.

aykevl commented on ports/nrf/mphalport.c:
```diff
@@ -77,3 +77,87 @@ void mp_hal_stdout_tx_strn_cooked(const char *str, mp_uint_t len) {
+            ".syntax unified\n"
+#endif
+        "1:\n"
+        " SUBS %0, %0, #1\n"
```
It may be better to use named registers, like `subs %[delay_us], %[delay_us], #1`. It avoids forcing the `us` argument at `r0` (where it is anyway). I would say it is also more platform independent but I guess that doesn't really matter for such code...
See: http://www.ethernut.de/en/documents/arm-inline-asm.html

aykevl commented on ports/nrf/boards/pca10040/nrfx_config.h:
```diff
@@ -0,0 +1,37 @@
+#define NRFX_TWI0_ENABLED 1
+#define NRFX_TWI1_ENABLED 1
+
+#endif // NRFX_CONFIG_H
```
I assume this file replaces nrf52_hal_conf.h?

glennrub committed:
ports/nrf/modules/machine/i2c: Alignment. Renaming print function param 'o' to 'self_in'

glennrub committed:
ports/nrf/spi: Updating SPI machine module to use nrfx drivers.

glennrub committed:
ports/nrf: Renaming modules/machine/rtc.c/.h to rtcounter.c/.h to not confuse the peripheral with Real-Time Clock:

glennrub committed:
ports/nrf: Updating various files after renaming machine module RTC to RTCounter.

glennrub committed:
ports/nrf: Renaming RTC to RTCounter in modmachine globals dict table. Also updating object type name to reflect new module name.

glennrub committed:
ports/nrf: Fixing leftovers after renaming rtc to rtcounter.

glennrub committed:
ports/nrf: Early untested adoption of nrfx_rtc in RTCounter. Untested.

aykevl commented on ports/nrf/modules/machine/i2c.c:
```diff
@@ -105,43 +108,52 @@ mp_obj_t machine_hard_i2c_make_new(const mp_obj_type_t *type, size_t n_args, siz
+    nrfx_err_t err_code = nrfx_twi_rx(self->p_twi, addr, dest, len);
+
+    if (err_code != NRFX_SUCCESS) {
+        return -err_code;
```
`readfrom` and `writeto` should return a MP_Esomething error (for example MP_EIO).

aykevl commented:
Some feedback:
* `MICROPY_KBD_EXCEPTION` doesn't work anymore
* I think nrfx_config.h should not be per-port, but as a global file enabling/disabling features as they're set in mpconfigboard.h (e.g. only enable SPI when `MICROPY_PY_MACHINE_HW_SPI` is enabled). This reduces redundant code.
* I2C appears to finally work, nice!

aykevl committed:
nrf/modules/machine/i2c: Improve keyword argument handling

glennrub committed:
ports/nrf/modules/temp: Updating Temp machine module to use nrfx defined hal nrf_temp.h. Moving logic of BLE stack awareness to machine module.

glennrub committed:
ports/nrf/boards/pca10040: Enable machine Temp module.

aykevl committed:
nrf/modules/machine/rtcounter: Remove magic constants.

glennrub committed:
ports/nrf: Adding base support for nrfx module logging. Adding option to disable logging of UART as it might log its own setup over UART while the peripheral is not yet set up. Logging of UART could make sense if other transport of log is used.

glennrub committed:
ports/nrf: updating nrfx_log.h with more correct parenthisis on macro grouping.

glennrub committed:
ports/nrf: Updating nrfx logging with configuration to disable logging of UART module. The pattern can be used to turn off other modules as well. However, for now UART is the only module locking itself by logging before the peripheral is configured. Logging is turned off by default, can be enabled in nrfx_config.h by setting NRFX_LOG_ENABLED=1.

glennrub committed:
ports/nrf/modules/random: Updating modrandom to use nrfx hal for rng. Not using nrfx-driver for this peripheral as its blocking mode would do the trick on RNG. Moving softdevice aware code from legacy hal to modrandom.c.

aykevl committed:
nrf: Enable Peripheral Resource Sharing.

This enables TWI and SPI to be enabled at the same time.

aykevl committed:
nrf/Makefile: Define MCU sub variant (e.g. NRF51822/NRF51422)

aykevl committed:
nrf: Port TIMER peripheral to nrfx HAL.

aykevl committed:
nrf/modules/machine/rtcounter: Update for nrfx HAL.

aykevl committed:
nrf/modules/machine/uart: Optimize UART module

For a nRF51, this results in a size reduction of:
.text: -68 bytes
.data: -56 bytes

aykevl committed:
nrf/modules/machine/uart: Don't use magic index numbers.

aykevl committed:
nrf/modules/machine/uart: Fix off-by-one error.

For nrf51:
.text: -40 bytes

aykevl committed:
nrf/mpconfigport: Avoid unnecessary root pointers.

This saves 92 bytes of RAM.

aykevl committed:
nrf/modules/machine/i2c: Reduce RAM consumption.

Reductions for the nrf51:
flash: -108 bytes
RAM:   -72 bytes

aykevl committed:
nrf: Support SoftDevice with nrfx HAL.

aykevl committed:
nrf: Add NVMC peripheral (microbitfs) support.

There is no support yet for a SoftDevice.

It also fixes a potentially serious bug in start_index generation.

aykevl committed:
nrf/modules/machine/spi: Optimize SPI peripheral.

nrf51:
text: -340 bytes
data: -72  bytes

nrf52:
text: -352 bytes
data: -108 bytes

aykevl commented:
Note to self, this is still TODO:

- [x] NVMC + SD - I think this is fixed
- [ ] IRQ driven UART (for KeyboardInterrupt)
- [x] nrfx_config.h global, for all boards
- [ ] GC root pointers for IRQ callbacks (at least Timer and RTCounter, maybe also Pin change interrupts) (will be done later, this was already an issue before this PR)
- [ ] TWI error codes are wrong

glennrub committed:
ports/nrf/microbit: Port microbit targets to nrfx HAL

Initial port of microbit modules to use nrfx HAL layer.
Tested display/image and modmusic on micro:bit to verify that
softpwm and ticker for nrf51 is working as expected.

Changing IRQ priority on timer to priority 2, as 1 might collide if
used side by side of SD110 BLE stack.

The patch reserves Timer1 peripheral compile time. This is not ideal
and should be resolved in seperate task.

aykevl committed:
nrf: Make nrfx_config.h universal for all boards.

aykevl committed:
nrf/modules/random: Forgot to commit header file.

glennrub committed:
nrf: Upmerge dev branch with master, resolving conflict in pin.c

aykevl committed:
nrf: Use SoftDevice API for flash access when built for SD

aykevl committed:
nrf/drivers/bluetooth: Remove legacy HAL driver includes.

These were not used anymore so can be removed.

glennrub committed:
nrf: Resolve merge conflicts in pca10040 mpconfigboard.h, keeping branch version of settings until all python modules uses nrfx drivers.

glennrub committed:
nrf/boards/microbit: Remove custom nrfx_config.h from microbit target, adding disablement of timer1 if softpwm is enabled.

glennrub committed:
nrf/modules/machine/pwm: Updating machine PWM module to use nrfx HAL driver.

examples/nrf52_pwm.py and examples/nrf52_servo.py tested on pca10040.

glennrub committed:
nrf/adc: Update ADC module to use nrfx

glennrub committed:
nrf: Removing hal folder and boards nrf5x_hal_conf.h headers.

glennrub committed:
nrf: delete duplicate files from device folder which can be located in nrfx/mdk.

glennrub committed:
nrf/Makefile: Fetch system files from nrfx.

Testing on each device sub-variant to figure out which system file to
use. Reason for this is that nrf52.c is actually defining nrf52832.

Removing NRF_DEFINES parameter setting the device in use into the
same sub-variant test, as NRF52 is unique to nrf52832 when using nrfx.
Without this exclusion of -DNRF52 in compilation for nrf52840, the
device will be interpreted as a nrf52, hence nrf52832.

Also, changing name on variable SRC_NRF_HAL to SRC_NRFX_HAL to
explicitly tell the origin of the file.

glennrub committed:
nrf: Updating device #ifdefs to be more open to non-nrf51 targets.

glennrub committed:
nrf/modules/machine/uart: Removing second instance of UART for nrf52840 as it only has one non-DMA variant.

glennrub committed:
nrf/device: Removing system files as these are now used from nrfx/mdk

glennrub committed:
nrf: Moving startup files in device one level up as there is no need for deep hierarchy.

glennrub committed:
Merge branch 'master' into move_hal_to_nrfx

glennrub committed:
nrf/examples: Updating ubluepy_temp after moving RTCounter to nrfx.

glennrub committed:
Merge branch 'master' into move_hal_to_nrfx

glennrub committed:
nrf/nrfx_glue: Adding direct NVIC access for S110 BLE stack

If SoftDevice s110 has not yet been initialized, the IRQ will not be forwarded to
the application using the sd_nvic* function calls. Hence, direct access to cmsi
nvic functions are used instead if SoftDevice is not enabled.

glennrub committed:
nrf/drivers/ticker: Setting IRQ priority 3 on Timer1

SoftDevice fails to initilize if Timer1 has been configured to priority
level 2 before enabling the SD. The timer is set to priority 1, higher than BLE
stack in order to provide better quality of music rendering when used with the
music module. This might be too high, time will show.

glennrub committed:
nrf: Use NRF52_SERIES defined in nrfx/mdk/nrf.h as define value when testing for both nrf52(832) and nrf52840 variants.

glennrub committed:
Merge branch 'master' into move_hal_to_nrfx

glennrub committed:
nrf/modules/machine/timer: Fixing defines excluding Timer1 if ticker/softpwm is used.

glennrub committed:
nrf: Switching import form mpconfigboard.h to mpconfigport.h in nrfx_config.h as mpconfigboard.h might define default values for defines not set by board specific header.

glennrub committed:
nrf/modules/machine/uart: Enable UART RX by default

Enable rx by default after intiialization of the peripheral.
Else, the nrfx driver will re-enable rx for each byte read
on uart REPL, clearing the EVENT_RXDRDY before second byte,
which again will make second byte get lost and read will get stuck.

This happens if the bytes are transmitted nrf(51) while still
processing the previous byte. Not seen on nrf52, but should
also become an issue at higher speeds.

This patch sets rx to always be enabled. Hence, not clearing the event
between read bytes, and it will be able to detect next byte recieved
upon finishing the first.

glennrub committed:
nrf/modules/machine/i2c: nrfx integration fixes

Increasing speed to 400K.

Returning Address NACK's as MP error code; MP_ENODEV.

Returning MP_ETIMEOUT on all other error codes from TWI nrfx driver
except the ANACK.

Enabling and disabling the TWI peripheral before and after each transaction.

glennrub committed:
nrf/examples: Updating ssd1306_mod.py to split framebuffer transfer into multiple chunks

glennrub committed:
nrf/modules/machine/i2c: Return MP_EIO error if Data NACK occurs.

glennrub commented:
**Legacy HAL:** 
```
   text	   data	    bss	    dec	    hex	filename
 134548	    256	  12972	 147776	  24140	build-microbit/firmware.elf
 120004	    184	  29320	 149508	  24804	build-pca10028/firmware.elf
 136300	    280	  41808	 178388	  2b8d4	build-pca10040/firmware.elf
 138308	    332	 140144	 278784	  44100	build-pca10056/firmware.elf
```
**NRFX:**
```
   text	   data	    bss	    dec	    hex	filename
 137388	    144	  13056	 150588	  24c3c	build-microbit/firmware.elf
 122864	     72	  29404	 152340	  25314	build-pca10028/firmware.elf
 140524	     72	  42004	 182600	  2c948	build-pca10040/firmware.elf
 141560	     72	 140284	 281916	  44d3c	build-pca10056/firmware.elf
```

aykevl commented on ports/nrf/drivers/flash.c:
```diff
@@ -0,0 +1,132 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
```
Looks like I didn't check the spelling here. "Micro Python" is the old form.

aykevl commented on ports/nrf/drivers/flash.h:
```diff
@@ -1,9 +1,9 @@
 /*
- * This file is part of the MicroPython project, http://micropython.org/
+ * This file is part of the Micro Python project, http://micropython.org/
```
Same here (though GitHub has inferred the wrong rename).

aykevl commented on ports/nrf/modules/machine/adc.c:
```diff
@@ -30,80 +30,120 @@
-    {{&machine_adc_type}, &ADCHandle5},
-    {{&machine_adc_type}, &ADCHandle6},
-    {{&machine_adc_type}, &ADCHandle7},
+STATIC machine_adc_obj_t machine_adc_obj[] = {
```
I think this one can be `const`.

aykevl commented on ports/nrf/modules/machine/adc.c:
```diff
@@ -112,14 +152,102 @@ STATIC mp_obj_t machine_adc_make_new(const mp_obj_type_t *type, size_t n_args, s
+{
+    uint8_t battery_level;
+
+    if (mvolts >= 3000) {
```
I see this algorithm was already in the nrf port, but I think it would be nice to explain it a bit. What are those magic numbers and where do they come from? For what battery technology is this?

glennrub replied:
Yeah, this is a blury API. Its for coincell, and depends on a diode drop in the calculation. I personally find this bonus API usefull when using nordic kits that can run coincell, but its not generic enough. I believe this can be improved in the future, or removed and let python handle the ADC measurement itself. If so, then there should also be an API parameter to select internal reference.

aykevl replied:
Maybe put the battery technology in the name? Something like `coincell_level_in_percent` would make it clear this is only for coin cells.

aykevl commented on ports/nrf/modules/machine/adc.c:
```diff
@@ -30,80 +30,120 @@
+    enum { ARG_pin };
     static const mp_arg_t allowed_args[] = {
-        { ARG_NEW_PIN,  MP_ARG_OBJ, {.u_obj = MP_OBJ_NEW_SMALL_INT(-1) } },
+        { ARG_pin,  MP_ARG_OBJ, {.u_obj = MP_OBJ_NEW_SMALL_INT(-1) } },
```
I think this should be `MP_QSTR_pin`.

glennrub replied:
This is even wrong as i can see it. For now, this is really the instance of the AIN. Resolving AIN based on Pin is not yet implemented. So, i suggest to rename this to "id" instead.

aykevl commented on ports/nrf/modules/machine/pin.c:
```diff
@@ -543,7 +551,7 @@ STATIC const mp_rom_map_elem_t pin_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_pull),    MP_ROM_PTR(&pin_pull_obj) },
     { MP_ROM_QSTR(MP_QSTR_af),      MP_ROM_PTR(&pin_af_obj) },
-    { MP_ROM_QSTR(MP_QSTR_irq),     MP_ROM_PTR(&pin_irq_obj) },
+//    { MP_ROM_QSTR(MP_QSTR_irq),     MP_ROM_PTR(&pin_irq_obj) },
```
Are pin interrupts not supported or simply not yet implemented?

glennrub replied:
It's somewhat implemented. However, what is of implementation, commented out, works. Pin interrupt should be changed to port events before activated, so that all pins can be used and not limited by number of GPIOTE channels.

aykevl commented on ports/nrf/modules/machine/pwm.c:
```diff
@@ -37,94 +37,81 @@
+    uint16_t       pulse_width;
+    uint16_t       period;
+    nrf_pwm_clk_t  freq;
+    pwm_mode_t mode;
```
Indentation

glennrub committed:
nrf: Updating git submodule and users to nrfx v1.0.0.

glennrub committed:
nrf: Addressing review comments.

aykevl commented on ports/nrf/modules/machine/pwm.c:
```diff
@@ -37,94 +37,81 @@
+#elif NRF52840
+    NRFX_PWM_INSTANCE(3),
+#else
+    NULL
```
It's a bit odd to put `NULL` here, but I guess since PWM isn't supported on the nrf51 anyway it doesn't really matter.

aykevl commented on ports/nrf/modules/machine/spi.c:
```diff
@@ -248,75 +227,85 @@ STATIC MP_DEFINE_CONST_DICT(machine_spi_locals_dict, machine_spi_locals_dict_tab
 
     if (baudrate <= 125000) {
-        self->spi->init.freq = HAL_SPI_FREQ_125_Kbps;
+        config.frequency = NRF_SPI_FREQ_125K;
```
I *think* we can make this code a lot smaller by calculating the frequency from the baudrate directly (something like `config.frequency = baudrate * 268`). But that can be done after this PR is merged, as it is not directly related to the nrfx changes.

aykevl commented on ports/nrf/modules/machine/spi.c:
```diff
@@ -248,75 +227,85 @@ STATIC MP_DEFINE_CONST_DICT(machine_spi_locals_dict, machine_spi_locals_dict_tab
 STATIC void machine_hard_spi_deinit(mp_obj_t self_in) {
-    machine_hard_spi_obj_t *self = self_in;
-    spi_deinit(self->spi);
+    // TODO
```
If `SPI.init` and `SPI.deinit` that call the above functions (`machine_hard_spi_init` and `machine_hard_spi_deinit`) don't actually do anything, why provide them? Maybe it's less confusing if they are (temporarily) left out.

aykevl commented on ports/nrf/modules/machine/spi.h:
```diff
@@ -26,12 +26,7 @@
-    mp_obj_base_t base;
-    SPI_HandleTypeDef *spi;
-} machine_hard_spi_obj_t;
+#include "nrf_spi.h"
```
I think this #include can be moved to spi.c itself.

glennrub committed:
nrf/modules/machine/adc: Update adc module to follow v1.0.0 nrfx API.

aykevl commented on ports/nrf/modules/machine/uart.c:
```diff
@@ -41,43 +42,30 @@
-#endif
+STATIC const machine_hard_uart_obj_t machine_hard_uart_obj[] = {
+    {{&machine_hard_uart_type}, .p_uart = &instance0},
 };
```
This misses UART1 support for the NRF52840.

aykevl commented:
I see there are a few header files (modules/machine/*.h) that include the nrfx headers. I think that's unnecessary in most cases - they should be included in the C files.

glennrub committed:
nrf/modules/machine/spi: Implement init and deinit functions

Extending SPI objects with a config member such that
configuration can be kept between new() and init().

Moving initialization done in new() to common init
function shared between the module functions.

If SPI is already configured, the SPI peripheral will
be uninitialized before initalized again.

Adding logic to handle initialization of polarity and
phase. As well, updating default speed to 1M from 500K.

glennrub committed:
nrf/modules/machine: Removing unused nrfx includes in machine module header files
